Bazel build rule for //test/cpp/util:grpc_cli

ivucica opened the pull request:
Cheers,

Here's a small patchset that makes it possible to build `grpc_cli` using Bazel.

First, to be able to use gflags's Bazel rules, this pulls in the current gflags master.

Then, some Bazel rules seem to have been missing references to some code. This became obvious once grpc_cli was being built.

Next, grpc_cli requires knowledge of the reflection service. Therefore, there's now a rule to build this library from the relevant proto. I've tagged it as `//visibility:public`; let me know if you'd like to restrict this for some reason.

Finally, there's a commit which adds a `cc_binary()` rule for building `grpc_cli` itself.

It could be a bit hacky as I've done this in my spare time; happy to add to the patchset (and perhaps squash resulting commits as required). If you'd prefer an internal CL, feel free to reach out to me internally and let me know where to contribute. I'll adapt the patches accordingly.

Tested only on amd64 Debian with Bazel 0.4.3 (latest stable). I have not tried to build with any other build system except Bazel.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

ivucica commented:
Jenkins is very chatty today.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

ivucica commented:
I've rebased to latest code and pulled back to use gflags 2.2.0 (which seems to have Bazel rules). I'm now trying to build this locally.

ivucica commented:
```
test/cpp/util/grpc_tool.cc:43:27: fatal error: gflags/gflags.h: No such file or directory
 #include <gflags/gflags.h>
                           ^
```

I think we will want to pull in newer gflags.

The latest gflags commit 
https://github.com/gflags/gflags/commit/30dbc81fb5ffdc98ea9b14b1918bfe4e8779b26e seems like it may be addressing this very issue. Maybe we can get @schuhschuh to cut a new release if that is the criteria? My original PR definitely pulled the latest commit and it worked.

y-zeng commented:
Jenkins: test this please

y-zeng commented:
Thanks for adding the bazel build rule for grpc_cli!
Could you please update the gflag version in [tools/run_tests/sanity/check_submodules.sh](../blob/master/tools/run_tests/sanity/check_submodules.sh), and add the copyright notice in [src/proto/grpc/reflection/v1alpha/BUILD](https://github.com/ivucica/grpc/blob/1b6b65fab009e3dd8b141192772d1c89245b8823/src/proto/grpc/reflection/v1alpha/BUILD) and [test/cpp/util/BUILD](https://github.com/ivucica/grpc/blob/1b6b65fab009e3dd8b141192772d1c89245b8823/test/cpp/util/BUILD).

ivucica commented:
Rebased and addressed. 

- Someone else created the reflection and util BUILDfiles, so I didn't have to add the copyright notice -- it's already there.
- I updated the submodule checker.

nicolasnoble commented:
Jenkins: test this please.

nicolasnoble commented:
Jenkins: this is ok to test.

ivucica commented:
A bunch of tests don't seem to have started; a few that did are failing in odd ways. What did I break? :)

y-zeng commented:
Jenkins: test this please.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

ivucica commented:
I have rebased, which got rid of another commit (no need for a change in gflags commit anymore).

Can we push this along?

ivucica commented:
I just spotted some additional changes which may no longer be necessary (specifically, submodules). Taking a look if I can revert them.

ivucica commented on test/cpp/util/BUILD:
```diff
@@ -34,8 +34,8 @@ licenses(["notice"])  # 3-clause BSD
     srcs = [],
-    deps = ["//:grpc++_unsecure"],
     linkshared = 1,
+    deps = ["//:grpc++_unsecure"],
```
This was done automatically when I saved the file on my workstation, I can undo it.

ivucica commented:
I am surprised at these test failures given that the PR is now literally one Bazel rule addition.

jtattermusch commented on test/cpp/util/BUILD:
```diff
@@ -104,5 +104,29 @@ cc_test(
+    name = "grpc_cli",
+    srcs = [
+        "cli_call.cc",
+        "cli_call.h",
```
headers should be separate?

jtattermusch replied:
https://github.com/grpc/grpc/blob/master/build.yaml#L1320

and 

https://github.com/grpc/grpc/blob/master/build.yaml#L3702

ivucica replied:
As discussed on hangouts: `cc_binary()` has no `hdrs`, and even in `cc_library()` it's valid to list in `srcs` various `.h` files that are not intended to be consumed by other rules.

I am doing some work on reorganizing this anyway, for possible testability and reuse. I may send it as a new PR.

jtattermusch commented:
Known test failures:

https://github.com/grpc/grpc/issues/10882
https://github.com/grpc/grpc/pull/10877
https://github.com/grpc/grpc/issues/10839
